Create new temporary directory/resource extension

jbduncan opened the pull request:
Proposed commit message:

```
${action} (${issues} / ${pull-request}) [max 70 characters]

${body} [max 70 characters per line]

${references}: ${issues}
PR: ${pull-request}
```

---
**PR checklist**

The following checklist shall help the PR's author, the reviewers and maintainers to ensure the quality of this project.
It is based on our contributors guidelines, especially the ["writing code" section](https://github.com/junit-pioneer/junit-pioneer/blob/main/CONTRIBUTING.md#writing-code).
It shall help to check for completion of the listed points.
If a point does not apply to the given PR's changes, the corresponding entry can be simply marked as done. 

Documentation (general)
* [x] There is documentation (Javadoc and site documentation; added or updated)
* [x] There is implementation information to describe _why_ a non-obvious source code / solution got implemented
* [x] Site documentation has its own `.adoc` file in the `docs` folder, e.g. `docs/report-entries.adoc`
* [x] Only one sentence per line (especially in `.adoc` files)
* [x] Javadoc uses formal style, while sites documentation may use informal style

Documentation (new extension)
* [x] The `docs/docs-nav.yml` navigation has an entry for the new extension
* [x] The `package-info.java` contains information about the new extension

Code
* [x] Code adheres to code style, naming conventions etc.
* [x] Successful tests cover all changes
* [x] There are checks which validate correct / false usage / configuration of a functionality and there are tests to verify those checks
* [x] Tests use [AssertJ](https://assertj.github.io/doc/) or our own [PioneerAssert](https://github.com/junit-pioneer/junit-pioneer/blob/main/CONTRIBUTING.md#assertions) (which are based on AssertJ)

Contributing
* [ ] A prepared commit message exists
* [ ] The list of contributions inside `README.md` mentions the new contribution (real name optional) 

---

I hereby agree to the terms of the [JUnit Pioneer Contributor License Agreement](https://github.com/junit-pioneer/junit-pioneer/blob/main/CONTRIBUTING.md#junit-pioneer-contributor-license-agreement).

jbduncan commented:
Note for self: I currently use `System::identityHashCode` to uniquely identify resources and resource factories, but this is [error-prone](https://github.com/plume-lib/plume-util/blob/ccdb37b4393bdbbc0087e3317b9a1d1b4827b78b/src/main/java/org/plumelib/util/UniqueIdMap.java) apparently, so use an atomic counter like `AtomicLong` instead.

Also consider searching junit-pioneer for similar usages of `(identity)HashCode` and `Atomic(Integer|Long)`, refactoring those usages out into a `UniqueIdMap` class, and raising it as a suggestion in a new PR.

nipafx commented on buildSrc/build.gradle:
```diff
@@ -3,24 +3,26 @@ import groovy.xml.XmlUtil
-							IssueNavigationLink {
-								option(name: "issueRegexp", value: '#(\\d+)')
-								option(name: "linkRegexp", value: 'https://github.com/junit-pioneer/junit-pioneer/issues/$1')
+			if (xmlFile.exists() && xmlFile.isFile()) {
```
You added this check, right? Can you let us know why it was necessary? If we need to change this file, we should so it in a separate change.

jbduncan replied:
I did indeed. It was necessary because when I imported this project into my IntelliJ IDEA, the file `.idea/vcs.xml` was absent for some bizarre reason. So whenever I tried to run any of the Gradle tasks, it threw an exception upon trying to parse the file.

That's a good shout, putting this in a separate change. Given I'm busy IRL, is this something that someone from the core team would be able to handle? If not, then that's fine, as I've just put it in my TODO list.

nipafx commented on src/main/java/org/junitpioneer/jupiter/TemporaryDirectory.java:
```diff
@@ -0,0 +1,80 @@
+
+			@Override
+			public Path get() throws Exception {
+				return (tempDir = Files.createTempDirectory(prefix));
```
Doesn't this reassign a new `tempDir` every time `get()` is called? Generally speaking, I'd prefer to have a final field, whose value gets assigned on creation.

jbduncan replied:
It does. I'm assumed that `Resource::get` will be called at most once by the extension and will never be called by user code.

I drove this code in a "minimalist" TDD fashion, so thanks for pointing out the flaw! It's now in the TODO list.

nipafx commented on src/main/java/org/junitpioneer/jupiter/ResourceManagerExtension.java:
```diff
@@ -0,0 +1,71 @@
+	@Override
+	public Object resolveParameter(ParameterContext parameterContext, ExtensionContext extensionContext)
+			throws ParameterResolutionException {
+		// TODO: Check that the parameter is annotated with at least @New or @Shared.
+		// TODO: Check that the parameter is not annotated with both @New and @Shared.
```
This was already checked in `supportsParameter`. We still need to find and interpret the annotations, but we don't need to verify proper use of them.

jbduncan replied:
Oh, good shout! I agree. They're now in the TODO list to remove.

nipafx commented on src/main/java/org/junitpioneer/jupiter/New.java:
```diff
@@ -0,0 +1,33 @@
+
+	Class<? extends ResourceFactory<?>> value();
+
+	String[] arguments() default {};
```
Can be an `Object` array, right? (So we can pass numbers as well.)

nipafx replied:
Never mind. 🙈 (`Object[]` is an invalid type for an annotation member.)

jbduncan replied:
> Never mind. :see_no_evil: (Object[] is an invalid type for an annotation member.)

Indeed. :)

It's a right shame, but I can't think of a better way.

nipafx commented on src/main/java/org/junitpioneer/jupiter/ResourceManagerExtension.java:
```diff
@@ -0,0 +1,71 @@
+				ReflectionSupport.newInstance(newResourceAnnotation.value(), (Object[]) newResourceAnnotation.arguments());
+		extensionContext
+				.getStore(
+					ExtensionContext.Namespace.create(ResourceManagerExtension.class, extensionContext.getUniqueId()))
```
I don't think we need `extensionContext.getUniqueId()` here - each extension context already has its own store.

jbduncan replied:
Ah, okay. I assume something like `Namespace.create(ResourceManagerExtension.class)` in a constant like `NAMESPACE` will do?

nipafx commented on src/main/java/org/junitpioneer/jupiter/ResourceManagerExtension.java:
```diff
@@ -0,0 +1,71 @@
+		extensionContext
+				.getStore(
+					ExtensionContext.Namespace.create(ResourceManagerExtension.class, extensionContext.getUniqueId()))
+				.put(System.identityHashCode(newResourceAnnotation) + "-resource-factory",
```
I understand the need for `identityHashCode` as follows:

* the store is a key-value store, so we need a key
* we need to make sure not to accidentally reuse the same key or already-registered resources would get kicked out again

Is that correct? If so, I see one other options besides `identityHashCode`:

* use a random `long` - chances of collisions are minuscule

jbduncan replied:
That's correct!

I actually made the same observation a few days ago, as I learned that `System::identityHashCode` [isn't generally a good way of IDing things](https://github.com/plume-lib/plume-util/blob/f7c1a3696ed26dd54a9659bef126a8bf4ea27d9e/src/main/java/org/plumelib/util/UniqueId.java#L9-L14) because when an object is garbage-collected, its memory location can be reused by a new object.

This may not apply in this case, but I think a `long` or `AtomicLong` would be best in case my assumption is ever invalidated.

It's already in the TODO list.

nipafx commented on src/main/java/org/junitpioneer/jupiter/ResourceManagerExtension.java:
```diff
@@ -0,0 +1,71 @@
+		extensionContext
+				.getStore(
+					ExtensionContext.Namespace.create(ResourceManagerExtension.class, extensionContext.getUniqueId()))
+				.put(System.identityHashCode(newResourceAnnotation) + "-resource",
```
As above, but one more option:

* ask the resource for a unique name - by default it could return its own `identityHashCode` (or the  random long), but certain resources may already be unique (e.g. file path for `Resource<Path>`)

jbduncan replied:
I think using a random `long`/`AtomicLong` is easiest to understand, so unless you have any objections, then that's the method I will try first. :)

nipafx commented on src/main/java/org/junitpioneer/jupiter/ResourceManagerExtension.java:
```diff
@@ -0,0 +1,71 @@
+								.orElse("unknown method")));
+		});
+		ResourceFactory<?> resourceFactory =
+				ReflectionSupport.newInstance(newResourceAnnotation.value(), (Object[]) newResourceAnnotation.arguments());
```
I know I said to pass these arguments to the constructor of the resource factory, but now I think that was wrong. I think it should be passed to `ResourceFactory::create`. That seems to make more sense.

jbduncan replied:
Oh, yep, that makes sense. It may make `Resource(Factory)` easier to understand. It's in the TODO list now.

nipafx commented on src/test/java/org/junitpioneer/internal/TestExtensionContext.java:
```diff
@@ -35,12 +35,12 @@ public TestExtensionContext(Class<?> testClass, Method testMethod) {
 	public Optional<Class<?>> getTestClass() {
-		return Optional.of(testClass);
+		return Optional.ofNullable(testClass);
 	}
 
 	@Override
 	public Optional<Method> getTestMethod() {
-		return Optional.of(testMethod);
+		return Optional.ofNullable(testMethod);
```
Can you explain why these changes are needed?

jbduncan replied:
@nipafx It's so that I can make a `TestExtensionContext` that has no test method.

This is to check an (admittedly extremely unlikely) edge case in `ResourceManagerExtension` where all hell has broken loose and:

1. `ResourceManagerExtension::supportsParameter` thinks a parameter has `@New`, but `::resolveParameter` disagrees.
2. And JUnit 5 itself gives `::resolveParameter` an `ExtensionContext` with an absent `getTestMethod()`.

My thinking was to ensure a helpful exception is thrown even in this case.

But I can see now that the change to `TestExtensionContext::getTestClass` is not needed. And regardless, this generally feels convoluted.

So I'm happy to revert this hunk and avoid testing this edge case if you'd like?

(For context, here's the [edge case](https://github.com/junit-pioneer/junit-pioneer/pull/491/files#diff-23a9e64b2f36fe2125ffc4a2aa13641e97c9c213cd3e0cffbc0a1b30aaa9e753R41) I talked about.)

nipafx commented on src/main/java/org/junitpioneer/jupiter/Shared.java:
```diff
@@ -0,0 +1,11 @@
+
+public @interface Shared {
+
+  Class<? extends ResourceFactory<?>> value();
```
I think using `value` as a name for the attribute makes the most sense when you can expect that it will be the only configured attribute in most use cases (because you can spare the `value = `). But since we always need the `name` as well, users would always have to write `value`, in which case I think we should use a more explicit name. What about `factory`?

jbduncan replied:
Yep, that makes sense! It's now in the TODO list.

nipafx reviewed (commented):
This looks good and is definitely going into the right direction. Keep at it. 😉 (And don't hesitate to let us know if you don't have the time or interest for it any more. We don't want to put pressure on you completing this.)

jbduncan commented:
@nipafx Wow, thank you for the detailed review! I'm still very interested in working on this, I'm just held back by IRL things again and I don't know when I'll be next free. I might be able to find time over the next few days to respond to the review.

Bukama commented:
@jbduncan Take your time - RL is more important than Pioneer!

Bukama commented:
Can you give us a short summary of the current status of the draft?

jbduncan commented:
@Bukama The only things left to do are as follows:

- Ensure that `@BeforeEach`/`@BeforeAll`/`@AfterEach`/`@AfterAll` methods annotated with `@Shared` are run sequentially.
- Raise some new issues to:
  - Ensure that the `TemporaryDirectory` resource factory implementation fixes all problems with JUnit 5's own `TempDir`, as described in the [issue OP](https://github.com/junit-pioneer/junit-pioneer/issues/348#issue-711396233).
  - Create a `TemporaryFile` resource factory.
- Write the documentation.

I intend to remove the draft status and open this PR for review when the first point is done.

nipafx commented:
@jbduncan No pressure, but #141/#541 is blocked by this. 😁 If you currently don't have the time to work on this, lets discuss how you can hand this over (IIRC, there are a bunch of changes on your machine that are not in the PR, yet).

jbduncan commented:
@nipafx Thanks for following up on this! I believe all the changes we worked on together are now on [GitHub](https://github.com/junit-pioneer/junit-pioneer/pull/491/commits/bab15506ed0d2e0346535aa98d849236d75d7525).

I saw https://github.com/junit-pioneer/junit-pioneer/issues/141 in my inbox recently, and I agree it looks it would benefit from my work. So I'm spending a bit of time this evening to fix the [last few bits](https://github.com/junit-pioneer/junit-pioneer/pull/491#issuecomment-986091281) I need to do before I can put this PR forward for review.

Be on the lookout for further news. :smile:

jbduncan commented:
@nipafx @Bukama This PR is _finally_ ready for review! :tada: 

Let me know what you think.

nipafx commented on src/test/java/org/junitpioneer/jupiter/resource/PathDeleterTests.java:
```diff
@@ -0,0 +1,83 @@
+
+class PathDeleterTests {
+
+	@Test
+	void deletesFile() throws IOException {
+		try (FileSystem fileSystem = Jimfs.newFileSystem(Configuration.unix())) {
+			Path file = fileSystem.getPath("file.txt");
+			Files.createFile(file);
+
+			Files.walkFileTree(file, PathDeleter.INSTANCE);
+
+			assertThat(file).doesNotExist();
+		}
+	}
```
Added tests to verify things actually get deleted. 😉

jbduncan replied:
Thanks!

These tests are not strictly needed because the deletion functionality in `PathDeleter` is already covered by `TemporaryDirectory`'s tests. However these new tests act as a nice sanity check in case the `TemporaryDirectory` tests are ever changed or removed, so I'm happy to keep these. :)

nipafx reviewed (commented):
Great PR, this going really well. I didn't review all of it, yet, but I made a bunch of small changes ☝🏾 and comments 👇🏾. Will have to continue another day.

jbduncan commented:
> Great PR, this going really well. I didn't review all of it, yet, but I made a bunch of small changes ☝🏾 and comments 👇🏾. Will have to continue another day.

@nipafx Thank you! And no worries, I understand, it's a very big PR.

I'll respond to your comments inline.

jbduncan commented on gradle.properties:
```diff
@@ -1,4 +1,4 @@
-junitVersion=5.7.2
+junitVersion=5.8.2
```
Thanks in advance, @nipafx. I was curious about this. :)

(For those not in the know, I upgraded the min JUnit 5 version to 5.8.0 at the start of this PR to allow the `@New` and `@Shared` annotations to be annotated with `@ExtendWith(ResourceExtension.class)`, reducing the amount of boilerplate code that users have to write when using this extension.)

jbduncan replied:
This thread will be resolved in https://github.com/junit-pioneer/junit-pioneer/pull/491/files#r808883191.

I'll resolve this thread in the meantime.

jbduncan commented on src/main/java/org/junitpioneer/jupiter/resource/ResourceExtension.java:
```diff
@@ -0,0 +1,368 @@
+import org.junit.platform.commons.support.AnnotationSupport;
+import org.junit.platform.commons.support.ReflectionSupport;
+
+class ResourceExtension implements ParameterResolver, InvocationInterceptor {
```
That makes sense, LGTM. :+1:

jbduncan commented on src/main/java/org/junitpioneer/jupiter/resource/ResourceExtension.java:
```diff
@@ -0,0 +1,368 @@
+		Optional<Shared> sharedAnnotation = parameterContext.findAnnotation(Shared.class);
+		if (sharedAnnotation.isPresent()) {
+			Parameter[] parameters = parameterContext.getDeclaringExecutable().getParameters();
+			// TODO: Replace root store with store of next outer class
+			ExtensionContext.Store rootStore = extensionContext.getRoot().getStore(NAMESPACE);
```
Perhaps you inferred this already, but the reason why I went for the root store was to allow for use cases like, "I have a database resource that I'd like to spin up once and share with my entire test suite".

I wonder if there's a way we can keep both your suggestion and what I originally had in mind. Perhaps using a `scope` attr, as you suggested, but with all these values instead: `Class`, `SourceFile`, `Package` (if possible), and `Global`.

I think one of these should be the default. I'm tempted to go with `Global`, but that doesn't seem the safest option. `Class` and `SourceFile` are the next most intuitive for me, but `Class` has the least scope, so I'd make that the default.

Thoughts?

jbduncan replied:
@nipafx ping ^

jbduncan replied:
@nipafx I'm hesitant to move forward with the last bits of implementation and then documentation until we agree on what the API for this should be. So I was wondering what your availability to review this is like?

nipafx replied:
Sorry for taking so long to reply. (I see, there's no shame emoji.) Lets go with `SourceFile` (default) and `Global`, initially.

I propose to leave out `Class` entirely because reusing the same name without sharing the resource in the same file sounds terribly unreadable to me. We can explore how to implement `Package` and `Module` (that would be neat) in separate PRs.

jbduncan commented on src/main/java/org/junitpioneer/jupiter/resource/ResourceExtension.java:
```diff
@@ -0,0 +1,368 @@
+		throw new ParameterResolutionException(errorMessage);
+	}
+
+	private <T> T checkType(Object resource, Class<T> type) {
+		if (!type.isInstance(resource)) {
+			String message = String.format("Parameter [%s] is not of the correct target type %s.", resource, type);
+			throw new ParameterResolutionException(message);
+		}
+		return type.cast(resource);
+	}
```
Agreed! Good catch. :+1:

jbduncan commented on src/main/java/org/junitpioneer/jupiter/resource/ResourceExtension.java:
```diff
@@ -0,0 +1,368 @@
+		}
+	}
+
+	private Object resolveShared(Shared sharedAnnotation, Parameter[] parameters, ExtensionContext.Store store) {
```
Nice method rename. :+1:

nipafx replied:
Thanks. :)

jbduncan commented on src/main/java/org/junitpioneer/jupiter/resource/ResourceExtension.java:
```diff
@@ -0,0 +1,368 @@
+		}
+	}
+
+	private ResourceWithLock<?> createResource(Shared sharedAnnotation, ResourceFactory<?> resourceFactory) {
+		try {
+			return new ResourceWithLock<>(
+				resourceFactory.create(unmodifiableList(asList(sharedAnnotation.arguments()))));
+		}
+		catch (Exception e) {
+			throw new UncheckedParameterResolutionException(new ParameterResolutionException(
+				"Unable to create a resource from `" + sharedAnnotation.factory() + "`", e));
+		}
+	}
```
Nice refactoring. :+1:

jbduncan commented on src/main/java/org/junitpioneer/jupiter/resource/ResourceExtension.java:
```diff
@@ -0,0 +1,368 @@
+
+			if (factoryKey(sharedAnnotation).equals(presentResourceFactoryName)
+					&& !sharedAnnotation.factory().equals(presentResourceFactory.getClass())) {
+				throw new ParameterResolutionException(
+					"Two or more parameters are annotated with @Shared annotations with the name \""
+							+ sharedAnnotation.name() + "\" but with different factory classes");
```
I see now that one of the tests is failing, I think because it's expecting this exception to be thrown but it's not.

I'll investigate when I next have the time.

jbduncan replied:
Failing test fixed.

jbduncan commented on src/main/java/org/junitpioneer/jupiter/resource/ResourceExtension.java:
```diff
@@ -0,0 +1,368 @@
+				.isPresent();
+	}
+
+	private long uniqueKey() {
```
@nipafx Thanks for renaming this from `newKey()` to `uniqueKey()`, makes more sense. :+1:

jbduncan commented on src/main/java/org/junitpioneer/jupiter/resource/ResourceExtension.java:
```diff
@@ -0,0 +1,368 @@
+		return runSequentially(invocation, invocationContext, extensionContext);
+	}
+
+	@Override
+	public void interceptBeforeAllMethod(Invocation<Void> invocation,
+			ReflectiveInvocationContext<Method> invocationContext, ExtensionContext extensionContext) throws Throwable {
+		runSequentially(invocation, invocationContext, extensionContext);
+	}
+
+	@Override
+	public void interceptAfterAllMethod(Invocation<Void> invocation,
+			ReflectiveInvocationContext<Method> invocationContext, ExtensionContext extensionContext) throws Throwable {
+		runSequentially(invocation, invocationContext, extensionContext);
+	}
+
+	@Override
+	public void interceptBeforeEachMethod(Invocation<Void> invocation,
+			ReflectiveInvocationContext<Method> invocationContext, ExtensionContext extensionContext) throws Throwable {
+		runSequentially(invocation, invocationContext, extensionContext);
+	}
+
+	@Override
+	public void interceptAfterEachMethod(Invocation<Void> invocation,
+			ReflectiveInvocationContext<Method> invocationContext, ExtensionContext extensionContext) throws Throwable {
+		runSequentially(invocation, invocationContext, extensionContext);
+	}
```
@nipafx I wasn't able to find a way to test these methods, so `ResourcesParallelismTests` doesn't cover them. Do you have any strong feelings about this?

nipafx replied:
Wouldn't it be enough to add lifecycle methods to some of the tests to see these methods executed?

jbduncan replied:
@nipafx Sorry for the delay! I found some time this evening to start properly reading through your messages and do more coding. (I expect to have more time this coming Friday.)

Oh, good observation! I've just pushed a new commit that addresses this: [`2e8a377`](https://github.com/junit-pioneer/junit-pioneer/tree/2e8a3770428881729a207f5b42a69bffd6c9d5d2).

nipafx replied:
Yesterday I noticed that we actually have a process for updating JUnit 5 minor versions - see #594.

nipafx replied:
So when this is ready to be merged, we quickly create a PR for #594, merge it, then update this PR and merge it.

jbduncan replied:
@nipafx Good to know that we have a process on upgrading JUnit 5 versions, thanks! I'll keep this thread unresolved until #594 is done. :)

nipafx replied:
Jupiter dependency was updated. ✅

jbduncan commented on src/main/java/org/junitpioneer/jupiter/resource/package-info.java:
```diff
@@ -0,0 +1,5 @@
+/**
+ * TODO
+ */
```
Oops, thanks for catching this. I'll cover this when I do the docs. :)

jbduncan commented on src/test/java/org/junitpioneer/jupiter/resource/DirTests.java:
```diff
@@ -0,0 +1,78 @@
+import org.junitpioneer.testkit.ExecutionResults;
+import org.junitpioneer.testkit.PioneerTestKit;
+
+class DirTests {
```
That makes sense, thanks! :+1:

jbduncan replied:
Also thank you for moving everything into its own `resource` package, I like how it encapsulates things a bit better. :+1:

nipafx commented on src/main/java/org/junitpioneer/jupiter/resource/ResourceExtension.java:
```diff
@@ -0,0 +1,367 @@
+			return invocation.proceed();
+		}
+		finally {
+			// for dining philosophers, "[t]he order in which each philosopher puts down the forks does not matter"
+			// (quote from Wikipedia)
+			locks.forEach(ReentrantLock::unlock);
```
I changed this from reverse unlocking because I'm pretty sure we don't need that.

jbduncan replied:
Oh excellent, thanks for making the fix!

nipafx commented on src/test/java/org/junitpioneer/jupiter/resource/ResourcesParallelismTests.java:
```diff
@@ -0,0 +1,287 @@
+	// this method is written to fail if it is executed at overlapping times in different threads
+	private static void failIfExecutedInParallel(String testName) throws InterruptedException {
+		boolean wasZero = COUNTER.compareAndSet(0, 1);
+		System.out.println(testName + ": wasZero = " + wasZero);
+		assertThat(wasZero).isTrue();
+		// wait for the next test to catch up and potentially fail
+		Thread.sleep(TIMEOUT_MILLIS);
+		boolean wasOne = COUNTER.compareAndSet(1, 2);
+		System.out.println(testName + ": wasOne = " + wasOne);
+		assertThat(wasOne).isTrue();
+		// wait for the last test to catch up and potentially fail
+		Thread.sleep(TIMEOUT_MILLIS);
+		boolean wasTwo = COUNTER.compareAndSet(2, 0);
+		System.out.println(testName + ": wasTwo = " + wasTwo);
```
The `System.out.println` appear weird to me. Are they needed?

jbduncan replied:
They're only there for debugging purposes. I get the impression that you'd prefer they be removed? I've tentatively removed them in commit [`4c0936d`](https://github.com/junit-pioneer/junit-pioneer/pull/491/commits/4c0936dde7462855fce94d725419fc24659e8c1e), but please do let me know if you'd like me to revert it.

nipafx replied:
Yes, I prefer them removed. Thanks!

nipafx reviewed (commented):
I know it's moving slowly (my bad), but it _is_ moving and in the right direction, too. 👍🏾 

Replied up there ☝🏾 and left two more comments.

Bukama commented:
JUnit 5.8.2 should be available on main branch now.

nipafx commented:
It just occurred to me, I don't think sharing and arguments work well together. If shared resource "Foo" is first referenced without arguments and later with one, then there's no way to fulfill the second request. We might just have to pull arguments from shared resources.

nipafx commented:
I'm not sure about the name `@Dir` because if the parameter has a domain-specific name like `customerDir` the parameter would be `@Dir Path customerDir` and at that point, there's nothing that indicates that it's a temporary directory, which I think is important to communicate.

The obvious solution is `@TempDir`, which clashes with JUnit 5's annotation of the same name, so that's not great.

nipafx commented:
More ideas that need to be spun out into individual issues:

* "passing" arguments from `@Dir` to `@New`
* create temp-dir specific variant of `@Shared`

nipafx commented:
@jbduncan This PR is very far advanced - any chance we can push it over the finish line in the next weeks? 😃

jbduncan commented:
@nipafx Yep, IRL things kept me away again, but I should be able to dedicate some time this weekend. :)

jbduncan commented:
@nipafx Apologies, I wasn't able to dedicate any time to this PR this weekend after all due to more IRL things. Not sure when I'll next have time, but I've tentatively pencilled this coming weekend down in my calendar.

jbduncan commented:
I rebased this PR over main earlier this afternoon (as I saw that the Asciidoc code snippets work was merged in recently), and then I worked on documenting how to create `@New` resources.

My next task is to document how to create `@Shared` ones.

(I'm also aware that there are CI failures due to build warnings and such. I'll fix those at some point!)

nipafx commented:
I fixed the small issued, so the build can be green. :)

jbduncan commented:
@nipafx Perfect, thank you! I've pencilled in some more time to work on the docs this weekend.

I'll get there. Eventually. :smile:

jbduncan commented:
> It just occurred to me, I don't think sharing and arguments work well together. If shared resource "Foo" is first referenced without arguments and later with one, then there's no way to fulfill the second request. We might just have to pull arguments from shared resources.

@nipafx I know we discussed this outside of GitHub in the past, but I just wanted to acknowledge this and say that I'll document this.

> I'm not sure about the name `@Dir` because if the parameter has a domain-specific name like `customerDir` the parameter would be `@Dir Path customerDir` and at that point, there's nothing that indicates that it's a temporary directory, which I think is important to communicate.
> 
> The obvious solution is `@TempDir`, which clashes with JUnit 5's annotation of the same name, so that's not great.

I see where you're coming from.

Given we used the name `@TempDir` in the old version of Temporary Directory, I don't have much of an emotional response against that name anymore. What do you think?

> More ideas that need to be spun out into individual issues:
>
> - "passing" arguments from @Dir to @New
> - create temp-dir specific variant of @Shared

Done! https://github.com/junit-pioneer/junit-pioneer/issues/648, https://github.com/junit-pioneer/junit-pioneer/issues/649

jbduncan commented:
@nipafx @Michael1993 @beatngu13 I am finally ready for a proper review! 🎉 

Sorry in advance for the enormous PR. I'm not sure how it can be split into easier-to-review parts, as everything seems to interconnected to me right now. But I could look into splitting out the `@Shared` work into its own PR, if it would make reviewing easier.

nipafx commented on src/test/java/org/junitpioneer/jupiter/resource/DirTests.java:
```diff
@@ -0,0 +1,58 @@
+import org.junitpioneer.testkit.ExecutionResults;
+import org.junitpioneer.testkit.PioneerTestKit;
+
+class DirTests {
```
New thought: To make this directory easier to navigate, we should prefix test class names with their "topic", i.e. "Resource" or "TemporaryDirectory" where that makes sense.

```suggestion
class TemporaryDirectoryDirTests {
```

jbduncan replied:
Sounds good to me! I'll accept the suggestion.

nipafx commented on src/test/java/org/junitpioneer/jupiter/resource/PathDeleterTests.java:
```diff
@@ -0,0 +1,83 @@
+
+import org.junit.jupiter.api.Test;
+
+class PathDeleterTests {
```
I just noticed that I myself added tests here, but before that I wanted to comment with:

> I understand we're having fun writing tests, but so much fun that we're testing other peoples' code? :) Or are you testing assumptions of how these APIs work?

😁

jbduncan replied:
I was test-driving my code the top-down "mockist" fashion, and when I eventually got to testing this deep in the call stack, I realised that I needed to test the "unhappy paths" of the parts of the code using `Files::walkFileTree`. Specifically, what happens if the temporary files were somehow deleted before `Files::walkFileTree` runs?

The easy solution is to use `Files::delete`, but that would throw throw `IOException`s if the files were deleted already, and I wanted to the code to silently do nothing instead. So I wanted to use `Files::deleteIfExists`. But how can one test that? The on-disk filesystem is inherently so non-deterministic that I couldn't think how to consistently delete a file created by `TemporaryDirectory` before it got to the `Files::walkFileTree` code.

So I split the code passed to `Files::walkFileTree` into its own unit, the `PathDeleter`, and tested that independently.

I hope this makes sense. :grin:

nipafx replied:
I'm an idiot, I just overlooked the use of `PathDeleter.INSTANCE`. 🤦🏾 Even after reading your reply, it took me a few minutes.

jbduncan replied:
No worries, my explanation was quite convoluted, so I don't blame you for taking a few minutes. :)

nipafx commented on src/test/java/org/junitpioneer/jupiter/resource/ResourcesParallelismTests.java:
```diff
@@ -0,0 +1,317 @@
+		@DisplayName("then the test factories do not run in parallel")
+		@Execution(SAME_THREAD)
+		@Test
+		void thenTestFactoriesDoNotRunInParallel() {
```
I don't think we need to test this as an extension doesn't care whether it's operating on a test or test factory, right?

jbduncan replied:
My thinking was that test factories could have a `@New` or `@Shared`-annotated parameters, just like normal tests, so I wanted to check that test factories annotated with `@Shared` executed sequentially too.

Is my thinking incorrect somewhere?

nipafx replied:
I got that but I thought it doesn't matter to our extension because we don't interact with that. But we very much do when we implement `InvocationInterceptor`, particularly `interceptTestMethod`, `interceptTestFactoryMethod`, `interceptDynamicTest`, and `interceptTestTemplateMethod`.

jbduncan replied:
Great, I'm glad I didn't miss anything here. Thanks, Nicolai. :)

nipafx commented on src/test/java/org/junitpioneer/jupiter/resource/ResourcesParallelismTests.java:
```diff
@@ -0,0 +1,317 @@
+		@DisplayName("then the test templates do not run in parallel")
+		@Execution(SAME_THREAD)
+		@Test
+		void thenTestTemplatesDoNotRunInParallel() {
```
Same as above but with "test or test template". :)

jbduncan replied:
Similar to my comment about test factories, my thinking was that test templates could have `@New` or `@Shared`-annotated parameters, just like normal tests, so I wanted to check that test templates annotated with `@Shared` executed sequentially too.

Is my thinking incorrect somewhere?

nipafx commented on src/test/java/org/junitpioneer/jupiter/resource/ResourcesParallelismTests.java:
```diff
@@ -0,0 +1,317 @@
+		@DisplayName("then the test class constructors do not run in parallel")
+		@Execution(SAME_THREAD)
+		@Test
+		void thenTestClassConstructorsDoNotRunInParallel() {
```
Same as above but with "method or constructor".

jbduncan replied:
Just like my thinking with https://github.com/junit-pioneer/junit-pioneer/pull/491#discussion_r923871071 and https://github.com/junit-pioneer/junit-pioneer/pull/491#discussion_r923871657, I thought that test constructors could run in parallel if users enabled JUnit 5's parallel mode, and if they were annotated with `@Shared`, then I didn't want anything non-deterministic to happen.

Is my thinking incorrect somewhere again?

nipafx commented on src/test/java/org/junitpioneer/jupiter/resource/ResourcesTests.java:
```diff
@@ -0,0 +1,543 @@
+import org.junitpioneer.testkit.PioneerTestKit;
+
+@DisplayName("Resources extension")
+class ResourcesTests {
```
I think we need to test the contract of the general-resource-portion of our extension(s). I will leave some comments below where I think that should be tested.

jbduncan replied:
Sounds good. I'll resolve this particular thread then and tackle the specific ones below.

nipafx commented on src/test/java/org/junitpioneer/jupiter/resource/ResourcesParallelismTests.java:
```diff
@@ -0,0 +1,317 @@
+		assertThat(wasTwo).isTrue();
+	}
+
+	// At time of writing, we couldn't find a way to test that @{Before, After}{All, Each}-annotated methods are not
```
Isn't this promised by Jupiter? Is there anything we can do (wrong) to make these tests fail?

jbduncan replied:
Excellent question. I have no idea. :)

I remember reasoning that if a test class was run in parallel, it's `@BeforeEach` and `@AfterEach` methods could be called many times concurrently. Nothing I've found in the JUnit 5 User Guide seems to disprove this, but I probably haven't looked hard enough.

But I'm aware that the tests below may not be pulling their weight. So I'd be happy to remove this comment and the tests below if you agree.

nipafx replied:
I looked into it and I think I understand what JUnit Platform does here. Each test (be it explicit, dynamic, or from a template) becomes a node in the test plan and that includes it's before/after behavior (see [this code](https://github.com/junit-team/junit5/blob/r5.9.0/junit-platform-engine/src/main/java/org/junit/platform/engine/support/hierarchical/NodeTestTask.java#L139-L161) in Platform's `NodeTestTask`). Parallelization happens on the level of test plan nodes, which means multiple nodes can be executed at the same time. That, in turn, means that before/after extensions for _different_ nodes can run in parallel.

It should be possible to test this behavior by writing a test class that has a long-running before/after method and at least two tests. Then the 2+ tests become 2+ test nodes that get executed in parallel and so the before/after methods should run in parallel.

@jbduncan: I'll leave this explanation here and continue with the review. I may have time to try this out for myself, in which case I'll let you know here to avoid duplicate work.

jbduncan replied:
@nipafx Wow, thank you for looking so deeply into this! I'll see if I can test the before/after behaviour now as suggested, but I'm away on holiday from noon UK time until Monday morning, so if you don't hear back from me by the end of today, then feel free to have a go yourself.

jbduncan replied:
I'm having a go at this now. I'll let you know how it goes. :)

jbduncan replied:
Hmm, it's proving harder than I imagined to get some working code for or to reason through. Probably too late at night for me. :P

@nipafx If I revisit this and find myself equally stuck, would you prefer I keep the existing tests or remove them?

nipafx commented on src/test/java/org/junitpioneer/jupiter/resource/ResourcesTests.java:
```diff
@@ -0,0 +1,543 @@
+				}
+
+			}
+
```
More tests:

* assert that for every `@New`, `ResourceFactory::create` is called
* assert that if `ResourceFactory::create` returns `null` or `Resource<null>` a proper exception is thrown (to avoid NPEs that would be confusing to users)

jbduncan replied:
Thanks for the feedback! Partially done: We now assert that if `ResourceFactory::create` returns `null` then a proper exception is thrown.

For my reference, the remaining tasks are:
- Assert that if `ResourceFactory::create` returns `Resource<null>` then a proper exception is thrown.
- Assert that for every `@New`, `ResourceFactory::create` is called.

jbduncan replied:
@nipafx This comment is now addressed! I trust you'll let me know if there's anything else that needs to be followed up here.

nipafx commented on src/test/java/org/junitpioneer/jupiter/resource/ResourcesTests.java:
```diff
@@ -0,0 +1,543 @@
+			}
+
+		}
+
```
* assert that for every `@Shared` with a new name, `ResourceFactory::create` is called
* assert that for every `@Shared` with a known name, `ResourceFactory::create` is not called

jbduncan replied:
@nipafx Sorry for the hiatus, I've just acted upon this feedback! See commit https://github.com/junit-pioneer/junit-pioneer/pull/491/commits/5ab5f3862ff22ec98f4f2d89e1e29694c38b3ab1.

nipafx commented on src/test/java/org/junitpioneer/jupiter/resource/ResourcesTests.java:
```diff
@@ -0,0 +1,543 @@
+
+	}
+
+	static class SingleTestMethodWithConflictingSharedTempDirParametersTestCases {
```
Shouldn't this also throw an exception?

```java
@Test
void test1(@Shared(factory = DummyResourceFactory.class, name = "some-name") String param) { }

@Test
void test2(@Shared(factory = OtherResourceFactory.class, name = "some-name") String param) { }
```

jbduncan replied:
@nipafx Oh, great catch! I'm glad you caught this, as it forced me to write a test that exposed a parallelism bug in the code, which I've fixed. See commit https://github.com/junit-pioneer/junit-pioneer/pull/491/commits/620900bbae5e16e38be643462d6db408cedcafc9. :D

nipafx reviewed (commented):
Great work! I forgot so much about the extension, I started reviewing from scratch with the documentation and pushed some changes (2daf566), then I checked many tests (not all; of the classes with tests in `org.junitpioneer.jupiter.resource` only `TemporaryDirectoryTests` still needs to be reviewed) before getting side-tracked. 😁

I will have more time for this next Thursday. I will either continue the review or make some of the changes myself. Let me know of you start working on some of these.

nipafx commented on src/test/java/org/junitpioneer/jupiter/resource/SingleTestMethodWithSourceFileScopedTempDirParameter.java:
```diff
@@ -0,0 +1,58 @@
+	static Path recordedPathForImplicit;
+	static Path recordedPathForExplicit;
+
+	static class TestCases {
```
I think a comment explaining why this file is written the way it is (i.e. with static inner classes that aren't `@Nested`) would make it easier to grok them.

jbduncan replied:
Thanks for the feedback! I agree, so I've moved this file into `TemporaryDirectoryTests.java` to make the association between the two files clearer, and I've added the following comment to every test case:

> 	// This test case class is static so that JUnit 5 doesn't pick it up
>	// automatically but our own tests can still explicitly find it and run it.

See commit https://github.com/junit-pioneer/junit-pioneer/pull/491/commits/436303a3a56615c99c8a286d0f580a4ba6b7c428.

nipafx replied:
When I made this comment...

```java
class SingleTestMethodWithSourceFileScopedTempDirParameter {
	static class TestCases {
		@Nested
		class Implicit {
			// tests...
		}
	}
}

... it was strange to see a top-level class without tests that had an inner class that was not `@Nested` (so it wouldn't get executed) and had a further inner class _with_ tests. After moving the code around, the setup is no longer strange at all - this arrangement of non-`@Nested` inner classes is pervasive throughout our tests, so I removed the comment again. 😬

jbduncan replied:
Aaah, now I understand. Thanks for shedding some light on this. I'm more than happy to see that the comments have been removed, then. Thanks!

nipafx commented on src/test/java/org/junitpioneer/jupiter/resource/SecondSingleTopLevelTestMethodWithSourceFileScopedTempDirParameter.java:
```diff
@@ -0,0 +1,54 @@
+import org.junit.jupiter.api.Nested;
+import org.junit.jupiter.api.Test;
+
+class SecondSingleTopLevelTestMethodWithSourceFileScopedTempDirParameter {
```
I think a comment explaining why this file is written the way it is would make it easier to grok them.

jbduncan replied:
Agreed, so similar to my [previous comment](https://github.com/junit-pioneer/junit-pioneer/pull/491#discussion_r953599706), I've moved this file into `TemporaryDirectoryTests.java` and commented all test cases to make it clearer why they're static.

nipafx commented on src/test/java/org/junitpioneer/jupiter/resource/TemporaryDirectoryTests.java:
```diff
@@ -0,0 +1,496 @@
+					.isNotEqualTo(SecondSingleTopLevelTestMethodWithSourceFileScopedTempDirParameter.recordedOuterPath);
+			assertThat(FirstSingleTopLevelTestMethodWithSourceFileScopedTempDirParameter.recordedPath)
+					.isNotEqualTo(SecondSingleTopLevelTestMethodWithSourceFileScopedTempDirParameter.recordedInnerPath);
+			assertThat(SecondSingleTopLevelTestMethodWithSourceFileScopedTempDirParameter.recordedOuterPath)
+					.isEqualTo(SecondSingleTopLevelTestMethodWithSourceFileScopedTempDirParameter.recordedInnerPath);
```
Isn't this a repetition of the previous tests, where you already verified that source-file scope works? This one is slightly different (with inner and outer class), but if you feel it's necessary to test that, I recommend to make that a separate test.

jbduncan replied:
Oh, good catch. Yeah, I can't remember why I went so far as to test with `@Nested` test classes here, so I've eliminated the duplication in commit https://github.com/junit-pioneer/junit-pioneer/pull/491/commits/fedd6d49a7720b0f366f232f613d3c94428d3b28.

nipafx commented on src/test/java/org/junitpioneer/jupiter/resource/TemporaryDirectoryTests.java:
```diff
@@ -0,0 +1,496 @@
+					.isNotEqualTo(SecondSingleTopLevelTestMethodWithSourceFileScopedTempDirParameter.recordedOuterPath);
+			assertThat(FirstSingleTopLevelTestMethodWithSourceFileScopedTempDirParameter.recordedPath)
+					.isNotEqualTo(SecondSingleTopLevelTestMethodWithSourceFileScopedTempDirParameter.recordedInnerPath);
+			assertThat(SecondSingleTopLevelTestMethodWithSourceFileScopedTempDirParameter.recordedOuterPath)
+					.isEqualTo(SecondSingleTopLevelTestMethodWithSourceFileScopedTempDirParameter.recordedInnerPath);
```
Isn't this a repetition of the previous tests, where you already verified that source-file scope works? This one is slightly different (with inner and outer class), but if you feel it's necessary to test that, I recommend to make that a separate test.

jbduncan replied:
Agreed. See comment https://github.com/junit-pioneer/junit-pioneer/pull/491#discussion_r953612409.

nipafx commented on src/test/java/org/junitpioneer/jupiter/resource/MorePaths.java:
```diff
@@ -0,0 +1,25 @@
+
+final class MorePaths {
+
+	static Path rootTempDir() {
```
It's a little dirty, but I think we can move this into `Assertions.java` (in this package) to get rid of this class. And you could make it a public static final field, initiated to the system property.

jbduncan replied:
That sounds sensible! Done: https://github.com/junit-pioneer/junit-pioneer/pull/491/commits/eeb123eb2367b0b29d4b596d40f6f73b537a818e

nipafx reviewed (commented):
Finished up the remaining changes in the `resource` test package. I think it will get more readable if we indeed prefix all class names that relate to the resources or the tempdir parts accordingly.

jbduncan commented:
@nipafx Thank you very much for such a thorough review! I've found the time to review some of your comments this evening, and I hope to have time to review the rest soon. Stay tuned.

jbduncan commented:
@nipafx I got around to addressing even more of your feedback this evening. See https://github.com/junit-pioneer/junit-pioneer/pull/491#discussion_r953079152 and https://github.com/junit-pioneer/junit-pioneer/pull/491#discussion_r953109683.

I don't suppose you've had time to look at my answers to your other questions, have you?
- [#1](https://github.com/junit-pioneer/junit-pioneer/pull/491#discussion_r923865970)
- [#2](https://github.com/junit-pioneer/junit-pioneer/pull/491#discussion_r923871071)
- [#3](https://github.com/junit-pioneer/junit-pioneer/pull/491#discussion_r923871657)
- [#4](https://github.com/junit-pioneer/junit-pioneer/pull/491#discussion_r923873595)
- [#5](https://github.com/junit-pioneer/junit-pioneer/pull/491#discussion_r923882876)

jbduncan commented:
@nipafx I believe I've now addressed all your current feedback.

If you find the time to review this further, please let me know if you have any questions about my unresolved responses to your comments. :)

nipafx commented:
![image](https://user-images.githubusercontent.com/6537432/189908669-9eb70223-8179-4353-b173-f9c096c71dec.png)
HOLYSHITITSHAPPENING

nipafx commented:
@jbduncan It's all done except:

* whatever you dislike about my commits: df96538, 6fa2bfa, 33803b5, 18a3a11, a943d63, c72d7f0, 6771094
* [testing before/after](https://github.com/junit-pioneer/junit-pioneer/pull/491#discussion_r969461025)

Michael1993 commented on src/test/java/org/junitpioneer/internal/AllElementsAreEqual.java:
```diff
@@ -0,0 +1,34 @@
+
+import org.assertj.core.api.Condition;
+
+public final class AllElementsAreEqual extends Condition<Iterable<?>> {
```
This is way too fancy. Delete this class and use something simpler, e.g.:
```java
List<Path> paths = TwoTestMethodsWithSharedSameNameTempDirParameterTestCases.recordedPaths;
assertThat(paths)
    .hasSize(2)
    .allSatisfy(path -> assertThat(path).isEqualTo(paths.get(0)))
    .allSatisfy(path -> assertThat(path).doesNotExist());
```

jbduncan replied:
Sounds reasonable. Fixed in https://github.com/junit-pioneer/junit-pioneer/pull/491/commits/3cbab7b21cd2aa7961f6b8e0814e2a4d08abd816.

Michael1993 commented on src/test/java/org/junitpioneer/internal/TestParameterContext.java:
```diff
@@ -0,0 +1,63 @@
+import org.junit.jupiter.api.extension.ParameterContext;
+import org.junit.platform.commons.support.AnnotationSupport;
+
+public class TestParameterContext implements ParameterContext {
```
Where is this used? Why is this here? I've deleted it on my local copy and the build is successful, so I think this should be deleted.

Michael1993 commented on src/main/java/org/junitpioneer/jupiter/resource/PathDeleter.java:
```diff
@@ -0,0 +1,41 @@
+import java.nio.file.SimpleFileVisitor;
+import java.nio.file.attribute.BasicFileAttributes;
+
+// visible for testing purposes
```
I don't really understand this comment... The only way this class could be *less* visible if it were private, which is not even permitted? Comment should be removed, probably.

jbduncan replied:
Oh, great catch! I'd forgotten that top-level classes can never be `private`, so this comment is redundant. Fixed in https://github.com/junit-pioneer/junit-pioneer/pull/491/commits/5f1b12e8b25f39f94fb46e8c7bf2d1d11dcea8b4.

Michael1993 commented on src/main/java/org/junitpioneer/jupiter/resource/PathDeleter.java:
```diff
@@ -0,0 +1,41 @@
+
+	static final PathDeleter INSTANCE = new PathDeleter();
+
+	// prevent instantiation
```
Very tiny thing, we usually put the comment _inside_ the constructor. See `PioneerPreconditions` for example.

jbduncan replied:
Okey dokey! Fixed in https://github.com/junit-pioneer/junit-pioneer/pull/491/commits/7b4c0f30507c0b01c7c9d7b9db47a65fada9c74e.

Michael1993 commented on src/main/java/org/junitpioneer/jupiter/resource/PathDeleter.java:
```diff
@@ -0,0 +1,41 @@
+// visible for testing purposes
+class PathDeleter extends SimpleFileVisitor<Path> {
+
+	static final PathDeleter INSTANCE = new PathDeleter();
```
Candidate for the [Enum singleton pattern](https://java-design-patterns.com/patterns/singleton/).

jbduncan replied:
I agree! But unfortunately, IntelliJ tells me that "No extends clause [is] allowed for enum". 😭

Michael1993 commented on src/test/java/org/junitpioneer/jupiter/resource/TemporaryDirectoryAssertions.java:
```diff
@@ -0,0 +1,45 @@
+
+	public static final Path ROOT_TEMP_DIR = Paths.get(System.getProperty("java.io.tmpdir"));
+
+	static void assertCanAddAndReadTextFile(Path tempDir) {
```
This not the way we do assertions. 😭 
Instead of `assertCanAddAndReadTextFile(path)` it should look like:
```java
assertThat(path)
    .canAddTextFile()
    .canReadTextFile();
```

Michael1993 replied:
But I'm confused - could this not be covered by `assertThat(path).isReadable().isWritable()`?

jbduncan replied:
Yeah, I thought that too, but I remember catching a subtle bug months ago that only unveiled by explicitly creating a file and writing to it. Admittedly I can't remember what this bug was though, so which would you prefer, `.can(Add|Read)TextFile()` or `is(Readable|Writable)()`?

(Assuming you'd prefer `.can(Add|Read)TextFile()`, I'll do some research now on how to do custom fluent assertions with AssertJ.)

jbduncan replied:
My work to translate the old assertion to `.can(Add|Read)TextFile()` has been pushed to commit [WIP - address some review feedback](https://github.com/junit-pioneer/junit-pioneer/pull/491/commits/20708c2219098b1e5646849d8e206bd0433dcb7d). Let me know what you think, such as if we should keep these assertions or stick with `.is(Readable|Writable)()`.

nipafx replied:
I [reviewed](https://github.com/junit-pioneer/junit-pioneer/pull/491#pullrequestreview-1171790003) this part.

Michael1993 commented on src/test/java/org/junitpioneer/jupiter/resource/TemporaryDirectoryAssertions.java:
```diff
@@ -0,0 +1,45 @@
+		}
+	}
+
+	static void assertReadableWriteableTemporaryDirectory(Path tempDir) {
```
As my previous comment, this should not look like this, in my opinion.
My recommendation:
```java
assertThat(tempDir)
    .isTemporaryDirectory()
    .isReadable()
    .isWritable();
```

Applies to `assertEmptyReadableWriteableTemporaryDirectory` as well.

jbduncan replied:
Sounds reasonable to me. Fixed in .

As part of that commit, I've elected to completely remove `TemporaryDirectoryAssertions` and move its sole remaining member, `ROOT_TEMP_DIR`, into its most frequent user, `TemporaryDirectoryTests`. For your information, the only other user of `ROOT_TEMP_DIR` is `TemporaryDirectoryDirTests`, which uses it only once.

Let me know if you think I should move `ROOT_TEMP_DIR` into its own file.

Michael1993 reviewed (changes requested):
I took a quick look at the PR and I think there is still a bit of room for polish.

The core code of the PR is super solid and looks great. It's a gargantuan effort and you deserve all the praise @jbduncan, thank you.

Michael1993 commented:
Jimfs is not in a module and does not define a module name in the manifest. Automatic modules are not supported by gradle. Not sure how to proceed.

Michael1993 commented:
[935be84](https://github.com/junit-pioneer/junit-pioneer/pull/491/commits/935be844ad3339f859231fd7cc5976c1bf78cc21) is just a mistake, please revert/remove if you can.

Michael1993 commented:
Hey @jbduncan sorry for pushing so much stuff on your branch, I'm just trying to make it work with our modular setup because I felt it would be unfair to push that burden on you.

jbduncan commented on src/main/java/org/junitpioneer/jupiter/resource/Scope.java:
```diff
@@ -0,0 +1,27 @@
+ * <p>For more details and examples, see
+ * <a href="https://junit-pioneer.org/docs/resources/" target="_top">the documentation on resources</a>.</p>
+ */
+public enum Scope {
```
@Michael1993 Thank you for taking a look at this PR! This is now done as of https://github.com/junit-pioneer/junit-pioneer/pull/491/commits/03bb47d50b6a9d1d0d20f7f2dd96ee16149e9a2c.

jbduncan commented on src/main/java/org/junitpioneer/jupiter/resource/TemporaryDirectory.java:
```diff
@@ -0,0 +1,75 @@
+
+	}
+
+	private static void deleteRecursively(Path tempDir) throws IOException {
```
Good catch! Fixed in https://github.com/junit-pioneer/junit-pioneer/pull/491/commits/81b8480eacef4c192590e9ab9f6b8f019e1cea6a.

jbduncan commented:
> Hey @jbduncan sorry for pushing so much stuff on your branch, I'm just trying to make it work with our modular setup because I felt it would be unfair to push that burden on you.

@Michael1993 No worries, I completely understand. My knowledge of modules is limited, so thank you very much for resolving this conflict with the main branch. :D

(I recently discovered the `org.gradlex.extra-java-module-info` plugin myself, so it's cool that we're using it here!)

Bukama commented:
Hey @jbduncan Can you give us an update about your progress and maybe a rough estimate of finishing it?

jbduncan commented:
@Bukama So sorry for the delayed response! Real life kept me busy once again, and I only just found the time to tackle this PR again this evening.

I'm currently stuck with ensuring that `@BeforeAll` and `@AfterAll` methods annotated with `@Shared` run sequentially. There seems to be a heisenbug that allows `@(Before|After)All` methods and `@Test` methods to run concurrently, which is eluding me. But I intend to look into it more tomorrow evening.

If you, @Michael1993 or @nipafx have time to review my latest changes, then any thoughts that might allow me to get unstuck would be greatly appreciated. The buggy code is most likely in the `ResourceExtension::intercept*` methods.

Michael1993 commented on src/main/java/org/junitpioneer/jupiter/resource/ResourceExtension.java:
```diff
@@ -323,7 +332,37 @@ public <T> T interceptTestClassConstructor(Invocation<T> invocation,
+		Set<Shared> sharedAnnotations = Stream
+				.concat(Stream.of(declaringClass), nestedClasses.stream())
+				.flatMap(clazz -> ReflectionSupport
+						.findMethods(clazz, unused -> true, HierarchyTraversalMode.TOP_DOWN)
```
Should be using [AnnotationSupport::findAnnotatedMethods](https://junit.org/junit5/docs/5.8.1/api/org.junit.platform.commons/org/junit/platform/commons/support/AnnotationSupport.html#findAnnotatedMethods(java.lang.Class,java.lang.Class,org.junit.platform.commons.support.HierarchyTraversalMode))

jbduncan replied:
I can see why you think that! But sadly it doesn't work in this case.

It can find methods like this:

```java
@Shared
@Test
public void someTest() { ... }
```

but not like this:

```java
@Test
public void someTest(@Shared parameter) { ... }
```

Hence the need to grab all methods, then filter over the ones with `@Shared`-annotated parameters.

jbduncan replied:
@Michael1993 I've found a way to eliminate the "heisenbug" in this area of the code, so this code no longer exists. Thank you very much for your thoughts anyway. :)

nipafx commented on src/test/java/org/junitpioneer/internal/PioneerPathAssert.java:
```diff
@@ -0,0 +1,78 @@
+		return this;
+	}
+
+	public PioneerPathAssert canReadTextFile() {
```
I prefer the name `canReadAndWriteFile`. This is a bit long, but the assertion _does_ write a file and that can fail on its own, so it needs to clear that the assertion does that. Same argument for no "Text" as above.

nipafx commented on src/test/java/org/junitpioneer/internal/PioneerPathAssert.java:
```diff
@@ -0,0 +1,78 @@
+	public PioneerPathAssert canReadTextFile() {
+		Path textFile = canAddTextFileInternal();
+
+		String expectedText = "some-text";
+		try {
+			Files.write(textFile, singletonList(expectedText));
+		}
+		catch (IOException e) {
+			throw failure("Cannot write to a text file");
+		}
```
I would add a parameter to the internal method that creates the file (`canAddTextFileInternal` for now), so it writes the text right away.

jbduncan replied:
Oh, good catch! But in addressing your other comments, I've found that `canAddTextFileInternal` is redundant, so I've removed it. Thank you very much, anyway.

See commit https://github.com/junit-pioneer/junit-pioneer/pull/491/commits/43ef4fd85c66202e76102cb3fdd93f59b5fdb1f7.

nipafx commented on src/test/java/org/junitpioneer/internal/PioneerPathAssert.java:
```diff
@@ -0,0 +1,78 @@
+		return this;
+	}
+
+	private Path canAddTextFileInternal() {
```
I prefer the name `canWriteFileWithText`, then the parameter as proposed above.

jbduncan replied:
That is indeed a better name, thanks! But in addressing your other comments, I've found that `canAddTextFileInternal` is redundant, so I've removed it. :)

See commit https://github.com/junit-pioneer/junit-pioneer/pull/491/commits/43ef4fd85c66202e76102cb3fdd93f59b5fdb1f7.

nipafx reviewed (commented):
Good catch by @Michael1993 on the assertions - this makes the change much better and provides a new building block for future tests. 👍🏾 Some comments on the PR below.

nipafx commented:
I have some time this week to look at Pioneer and this is at the top of my list, so I will see what can be done.

By the way, we [failed](https://twitter.com/nipafx/status/1589913545420177409) you @jbduncan - this PR grew entirely out of control. But you've kept at it. True hero! 🦸🏾‍♂️

nipafx commented on src/test/java/org/junitpioneer/internal/PioneerPathAssert.java:
```diff
@@ -0,0 +1,78 @@
+
+public class PioneerPathAssert extends PathAssert {
+
+	public static PioneerPathAssert assertThatPath(Path actual) {
+		return new PioneerPathAssert(actual);
+	}
```
Oh, `PioneerAssert` is actually just more than just an entry point. Will have to revisit that proposal tomorrow.

jbduncan replied:
Okay, I'm happy to wait for a consensus on this. :)

jbduncan commented:
@nipafx Thank you for your latest round of comments, as well as for your kind words. True, this PR's scope was _much_ larger than anticipated, but it's been partly on me too, as I've taken a number of long breaks from working on it. :)

Your review and @Michael1993's have been noted, but I've just had an idea on how to move forward with the heisenbug I mentioned last time, so I'll focus some of my attention on that first.

jbduncan commented:
On the bright side, this PR has taught me loads about JUnit Jupiter extensions and how to debug parallel code!

jbduncan commented:
@nipafx @Michael1993 D'oh! There was no heisenbug, my brain just failed me.

I was accidentally testing in this scenario that the _`@Test`s_ were running sequentially.

```java
class TestCases {

	@BeforeEach
	void setup1(@Shared(factory = Factory.class, name = "a") Object a,
			@Shared(factory = Factory.class, name = "b") Object b) {}

	@BeforeEach
	void setup2(@Shared(factory = Factory.class, name = "b") Object b,
			@Shared(factory = Factory.class, name = "c") Object c) {}

	@BeforeEach
	void setup3(@Shared(factory = Factory.class, name = "c") Object c,
			@Shared(factory = Factory.class, name = "a") Object a) {}

	@Test
	void test1() {
	}

	@Test
	void test2() {
	}

	@Test
	void test3() {
	}

}

```

I should have been testing the `@BeforeEach` methods instead, because they're the ones annotated with `@Shared`. :sweat_smile: :man_facepalming:

So it led to me down the wrong path when testing `@BeforeAll`. Thus the "heisenbug" has been eliminated.

Lesson learned! Double check your logic when working with parallel code.

jbduncan commented on src/test/java/org/junitpioneer/internal/PioneerPathAssert.java:
```diff
@@ -0,0 +1,78 @@
+ * http://www.eclipse.org/legal/epl-v20.html
+ */
+
+package org.junitpioneer.internal;
```
Ahh, gotcha. Not sure how I missed all those other `*Assert` classes there!

Fixed in commit https://github.com/junit-pioneer/junit-pioneer/pull/491/commits/9788022d893a4d389f2443c645cc49c78984e998.

jbduncan commented on src/test/java/org/junitpioneer/internal/PioneerPathAssert.java:
```diff
@@ -0,0 +1,78 @@
+		super(path);
+	}
+
+	public PioneerPathAssert canAddTextFile() {
```
Those are some good points. I'm convinced!

But in addressing your other comments, I found that this assertion is redundant, as every usage was looking like this:

```java
assertThatPath(tempDir).canWriteFile().canReadAndWriteFile();
```

So I've removed it.

See commit https://github.com/junit-pioneer/junit-pioneer/pull/491/commits/43ef4fd85c66202e76102cb3fdd93f59b5fdb1f7.

jbduncan commented:
I'm ready for the next round of reviews! (Hopefully the final one this time, fingers crossed. :crossed_fingers:)

nipafx reviewed (commented):
Found two new TODOs.

nipafx commented:
One more TODO: All new annotations need a `@since` attribute in their Javadoc. That should be the last thing we do to make sure we pick the right version.

jbduncan commented:
Ah, it's just occurred to me that by simplifying the `intercept*` logic for lifecycle methods like `@BeforeEach` and `@BeforeAll`, if multiple parallel tests try to mutate a shared resource injected into e.g. a `@BeforeAll` method, then race conditions could happen.

However, this would only ever be a problem if users tried to save shared resources in fields. So as another TODO for myself, I'll add a paragraph to the docs to warn users against this.

As a longer term solution, I'll raise an issue later to consider running all tests in a test class sequentially if that class has a lifecycle method injected with a shared resource.

nipafx commented on src/test/java/org/junitpioneer/testkit/assertion/PioneerAssert.java:
```diff
@@ -10,311 +10,28 @@
-	private Events getProperEvents() {
-		return this.test ? actual.testEvents() : actual.containerEvents();
-	}
+/**
+ * Entry point to all JUnit Pioneer assertions.
+ */
+public class PioneerAssert {
```
Done.

jbduncan commented:
A warning about not saving resources in fields has just been added in commit https://github.com/junit-pioneer/junit-pioneer/pull/491/commits/25a412f054b7ce53837826500a68074699da7cd1.

jbduncan commented on src/main/java/org/junitpioneer/jupiter/resource/ResourceExtension.java:
```diff
@@ -233,6 +234,8 @@ private void throwIfHasAnnotationWithSameNameButDifferentType(ExtensionContext.S
 
 	private void throwIfMultipleParametersHaveExactAnnotation(Parameter[] parameters, Shared sharedAnnotation) {
+		// TODO: throw if multiple parameters have @Shared annotations with not just the same
+		//       name and factory type but the same scope too?
```
I've decided that this TODO doesn't make much sense, as this method already checks that no two `@Shared` annotations can have the same name and factory type anyway, so it's redundant to check the scope too.

So this TODO is now removed, as of https://github.com/junit-pioneer/junit-pioneer/pull/491/commits/8c8d0033701e2079016f9375738e00b6558ff2db.

jbduncan commented on src/main/java/org/junitpioneer/jupiter/resource/ResourceExtension.java:
```diff
@@ -209,6 +209,7 @@ private Resource<?> newResource(Object newOrSharedAnnotation, ResourceFactory<?>
 
 	private void throwIfHasAnnotationWithSameNameButDifferentType(ExtensionContext.Store store,
 			Shared sharedAnnotation) {
+		// TODO: throw if has annotation with same name but different scope
```
I've just addressed this TODO in commit https://github.com/junit-pioneer/junit-pioneer/pull/491/commits/8c8d0033701e2079016f9375738e00b6558ff2db. :)

jbduncan commented:
Would you like me to fix the merge conflict between this branch and the main branch, team?

jbduncan commented:
I've also raised a [new issue](https://github.com/junit-pioneer/junit-pioneer/issues/685) for the parallelism problem I mentioned in https://github.com/junit-pioneer/junit-pioneer/pull/491#issuecomment-1310106521.

jbduncan commented:
@nipafx I've just addressed your comment about the missing `@since` attributes in commit https://github.com/junit-pioneer/junit-pioneer/pull/491/commits/ec99827bf5e20384bbb09235bfd7d3134f90781f. I've also fleshed out the JavaDocs a bit in that commit as well as commit https://github.com/junit-pioneer/junit-pioneer/pull/491/commits/e9ae08446e29e975b8e3fb1fd4d4ff7b1962ce58.

jbduncan commented:
We're so close, I can almost taste it! ✨

nipafx commented:
It happened! 🤯

TWiStErRob commented:
Whaaaat? 1.5 years in the making 👏🏻 
![image](https://user-images.githubusercontent.com/2906988/201693644-dcc645ca-e575-4d90-b319-e82608f53dce.png)

nipafx commented:
Thank you so much for your infinite patience, @jbduncan! 🙏🏾 We finally made it.

jbduncan commented:
@nipafx You're very welcome, and thank you and everyone else for all your guidance! What a ride and learning experience it's been. 😊

jbduncan commented:
I'll keep an eye out for future bug reports and suggestions with this extension, so that if I have the time, I can address then as soon as possible.
